### PR TITLE
[DeviceMesh] Rename _device_mesh.py to device_mesh.py to prepare for beta (#114710)

### PR DESCRIPTION
**Pull request by wz337**
Summary:

Same content of changes as https://github.com/pytorch/pytorch/pull/114710

Rename _device_mesh.py to device_mesh.py, update all callsites, adds documentation.
ghstack-source-id: 208980207
exported-using-ghexport

Test Plan: CI.

Reviewed By: wanchaol

Differential Revision: D51629761

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @kiukchung @d4l3k @lucasllc

### CODE REVIEW

**Review by wanchaol — APPROVED** [review]

**Review by fduwjj — APPROVED** [review]
congrats for going beta.

**Comment by wz337** [comment]
> congrats for going beta.

Finger crossed that I could get this landed by the end of today to make the release cut. T_T

**Review by fegin — APPROVED** [review]
Good luck for the release :)

**Comment by malfet** [comment]
This broke public bindings tests, isn't it?

**Comment by malfet** [comment]
@wanchaol let's see if it fixes is, if not I'll take care of re-landing/cherry picking the change

**Comment by huydhn** [comment]
> And on PR as well: https://github.com/pytorch/pytorch/actions/runs/7068084522/job/19242309778 (cc: @huydhn do you know why it was classified as flaky?)

Dr.CI wrongly classified this as flaky because it found the same error happening on a different PR at https://github.com/pytorch/pytorch/pull/114710.  The two were considered unrelated because they were committed under 2 different emails `irisz at meta.com` and `wz337 at cornell.edu` (logic added by https://github.com/pytorch/test-infra/pull/4688).  I will add a fix to get the actual GitHub username instead to fix the issue.

**Comment by malfet** [comment]
image

Failure is indeed related, let me check why...

**Comment by malfet** [comment]
Before this PR on MacOS: `python3 -c "import torch.distributed._composable"`, returned `ModuleNotFoundError`, which is expected behavior, after it fails with `AttributeError`

File: torch/distributed/__init__.py [hunk 1]
```diff
@@ -92,6 +92,7 @@ def breakpoint(rank: int = 0):
         )
 
     from .distributed_c10d import *  # noqa: F403
+    from .device_mesh import *  # noqa: F403
```
Comment (huydhn) [inline torch/distributed/__init__.py:95]: It looks like the `test_correct_module_names` failure is due to the fact that `device_mesh.py` is graduated to become a public module, thus is now subjected to the test.

One way to fix this expected issue from what I see is to add a stub like `_ProcessGroupStub` for `torch.distributed.ProcessGroup`, for example:

```
class _LoggerStub:
    pass
sys.modules["torch.distributed"].Logger = _LoggerStub  # type: ignore[attr-defined]
```

The comment on that part mentions that this is the expected workaround to satisfy the test?

**Comment by wz337** [comment]
Close this one and replace with https://github.com/pytorch/pytorch/pull/115193

**Pull request closed without merging**